guides/installing-python.md

```diff
@@ -0,0 +1,147 @@
+- [Setting the global Python versions](#setting-the-global-python-versions)
+- [Using local Python versions](#using-local-python-versions)
+- [Upgrading Python versions](#upgrading-python-versions)
+- [Further reading](#further-reading)
+
+## Migrating from Homebrew-installed Python
+
+**Note**: Any virtualenvs that you have prior to this process will have to be recreated to use Python from pyenv.
+
+### Remove Homebrew-installed virtualenvwrapper
+
+Before installing and configuring pyenv, uninstall virtualenvwrapper if it  is installed via Homebrew:
+
+```shell
+brew uninstall virtualenvwrapper
+```
+
+Then remove the sourcing of `virtualenvwrapper.sh` and related variables from your shell's dotfiles (`.profile`, `.bashrc` for bash users, `.zshrc` for zsh users):
+
+```shell
+export VIRTUALENVWRAPPER_PYTHON=$HOME/homebrew/bin/python
+export VIRTUALENVWRAPPER_VIRTUALENV=$HOME/homebrew/bin/virtualenv
+source $HOME/homebrew/bin/virtualenvwrapper.sh
+```
+
+### Optional: remove Homebrew-installed Python and variables
+
+Before installing and configuring pyenv, it is a good idea to uninstall all old versions of Python and virtualenvwrapper that might be installed via Homebrew, and remove any varaibles from your shell's dotfiles (`.profile`, `.bashrc` for bash users, `.zshrc` for zsh users):
+
+```shell
+brew uninstall python
+brew cleanup python
+brew uninstall virtualenvwrapper
+```
+You should also remove any variables that reference Homebrew's Python from your shell's dotfiles (`.profile`, `.bashrc` for bash users, `.zshrc` for zsh users), for example:
+
+```shell
+export PYTHONPATH=$HOME/homebrew/bin/python
+```
+
+If you do not wish to uninstall Homebrew's Python, or if you have Homebrew-installed packages that depend on Python-installed Homebrew, simply ensure that Homebrew's Python either does not exist in your shell's `PATH` variable, or that it is added to `PATH` *before* the pyenv configuration below.
+
+## Installing pyenv
+
+To [install pyenv on MacOS](https://github.com/pyenv/pyenv#homebrew-on-macos), use Homebrew:
+
+```shell
+brew update
+brew install pyenv
+brew install pyenv-virtualenv
+brew install pyenv-virtualenvwrapper
+```
+
+**Note:** pyenv-virtualenvwrapper is not strictly required, depending on how you wish to manage your virtualenvs. See the [notes in the pyenv-virtualewrapper README](https://github.com/pyenv/pyenv-virtualenvwrapper#pyenv-virtualenvwrapper). Some of our documentation assumes `virtualenvwrapper` exists, so this guide installs it. You will see a deprecation warning when using `mkvirtualenv`:
+
+```
+WARNING: the pyenv script is deprecated in favour of `python3.6 -m venv`
+```
+
+Until we switch to using Python 3 and the `venv` module full-time, this can be safely ignored.
+
+## Configuring pyenv
+
+Per the [pyenv documentation](https://github.com/pyenv/pyenv#basic-github-checkout), the follow environment variables need to be added to your shell dotfiles (`.bashrc` for bash users, `.zshrc` for zsh users).
```

Comment:
Previously, the doc has been noting `.profile` as well for bash users - should standardize.

Reply:
@higs4281 had some issues using `.profile` — it would be good to know which works for whom and try to figure out why.

Reply:
I didn't have trouble with `.profile` but I did have trouble if the virtualenvwrapper bits were in `.bash_profile` instead of `.bashrc`

Sequence-wise, `.profile` should behave the same as `.bash_profile`, so either way, for bash users, I think `.bashrc` is a good recommendation.

Reply:
Other note is on phrasing... theoretically `.bash_profile` is the bash-specific `.profile`. `.profile` itself should be used by any bourne-compatible shell. That's probably not a detail worth getting stuck on, but as a non-bash user I get twitchy.

Above I used:

> (`.profile`, `.bashrc` for bash users, `.zshrc` for zsh users)

Which kinda skirts around this by using the ambiguity of English.

Reply:
Reading further, it seems like bash **doesn't** use `.bashrc` if the shell is interactive *and* a login shell. So, `.profile` or `.bash_profile` is probably better. Would be interested in getting @higs4281 to test.

Reply:
Potentially useful article cited at the top of our standard `.bash_profile`: http://www.joshstaiger.org/archives/2005/07/bash_profile_vs.html

Reply:
In my testing, having virtualenvwrapper declarations in `.bash_profile` threw errors and moving them to `.bashrc` fixed it, at least in Mac OS. I'm not clear on what to re-test.

Reply:
For now I duplicated the line:

> (.profile, .bashrc for bash users, .zshrc for zsh users)

From earlier in the file. That seems like it should satisfy the cases above?

Reply:
Also, are we actually talking about `.bash_profile` in every instance, or do some people actually have a `.profile`?

Reply:
Nevermind, I see the comments above that I glossed over before.

Reply:
![shells](https://media.giphy.com/media/CnTDrbGaPxNGE/giphy.gif)